### DISCUSSION TITLE
Inconsistent return types of random_generator functions in cpu and gpu

### DISCUSSION THREAD

**Discussion by ChaiBapchya** [apache/incubator-mxnet#13187] [Q&A]
The file - `/incubator-mxnet/include/mxnet/random_generator.h` includes CPU and GPU implementation of random number generator functions.

Close look at `cpu` implementation
```
MSHADOW_XINLINE FType uniform() {...}
MSHADOW_XINLINE FType normal() {...}
```

While in case of `gpu`
```
MSHADOW_FORCE_INLINE __device__ float uniform() {...}
MSHADOW_FORCE_INLINE __device__ float normal() {...}
```

Why is it hardcoded in gpu? Shouldn't it be generic?

**Comment by leleamol** [comment]
Hello @ChaiBapchya 
For questions, please submit on MXNet discussion forum (https://discuss.mxnet.io), where it will get a wider audience and allow other to learn as well.
 I'm closing this issue now in favor of the discussion forum issue you will file, please feel free to re-open if closed in error.
 Thanks!

@mxnet-label-bot add [Question]

**Comment by leleamol** [comment]
@mxnet-label-bot add [Question]

**Comment by asitstands** [comment]
There is a template specialization for double in gpu. Generic implementation without specialization is not so simple because the underlying curand API  is not generic. It requires an implementation of a new generic random number generator. Actually I have tried to address some issues in random number generation by implementing a new generator common for cpu and gpu (https://github.com/dmlc/mshadow/pull/338), but no one might be interested in the proposal :grin:
